addons/wowl/static/src/services/action_manager.ts

```diff
@@ -0,0 +1,128 @@
+import type { OdooEnv } from "./../env";
+import { Type , Odoo } from "./../types";
+import { Service } from "./../services";
+
+
+interface dispatchParams {
+  [key: string]: any,
+}
+
+//type finalizeParams = 'commit' | 'rollback';
+
+interface ActionManager {
+  get(property: string): any,
+  dispatch(method: string, params: dispatchParams): void,
+}
+
+interface Slot {
+  active: boolean,
+}
+
+interface ControllerSlot extends Slot {
+  stack: string[],
+}
+
+interface AsyncSlot<T extends Slot> {
+  name: string,
+  Component: Type<Component>,
+  current: T,
+  pending?: T,
+  do(cmd: string, args: any): any
+  get(property: string, args: any): any
+}
+
+class GenericAction extends Component<{}, OdooEnv> {
+  static template = tags.xml`
+  <div>
+    <p t-esc="controller.display" />
+    <p>all</p>
+    <t t-foreach="getAllControllers()" t-as="ct">
+      <span t-esc="ct.display" />
+    </t>
+  </div>
+  `;
+  get controller() {
+    return (this.props as any).slot.get('controller') || {};
+  }
+  getAllControllers() {
+    return (this.props as any).slot.get('allControllers') || [];
+  }
+}
+
+function makeActionManager(env: OdooEnv): ActionManager {
+  const main: AsyncSlot<ControllerSlot> = {
+    name: 'main',
+    Component: GenericAction,
+    current: {
+      active: false,
+      stack: [],
+    },
+    do(cmd: string, args: any) {
+      const stack = this.pending ? this.pending.stack : this.current.stack.slice();
+      this.pending = this.pending || {
+        active: true,
+        stack
+      };
+
+      switch (cmd) {
+        case "push":
+          stack.push(args.controller);
+          break;
+        case "commit":
+          if (this.pending) {
+            this.current = this.pending;
+          }
+          this.pending = undefined;
+      }
+    },
+    get(property: string) {
+      const stack = this.pending ? this.pending.stack : this.current.stack;
+      switch (property) {
+        case "controller":
+          return stack[stack.length-1];
+        case "allControllers":
+          return stack;
+      }
+    }
+  };
+  function dispatch(cmd: string, args:any) {
+    switch (cmd) {
+      case "doAction":
+        doAction(args);
+        break;
+      case "slotDo":
+        slotDo(args);
+        break;
+    }
+    env.bus.trigger('actionManager:update');
+  }
+  function slotDo(args: any) {
+    main.do(args.method, args.args);
+  }
+  function doAction(args: any) {
+    const controller = {
+      display: args.display,
+    };
+    dispatch('slotDo', {name: 'main', method: 'push', args: {controller}});
+  }
+  env.bus.on('actionManager:finalize', null, finalize);
+  function finalize() {
+    main.do('commit', {});
+  }
+  return {
+    get(property: string) {
+      switch (property) {
+        case "slots":
+          return [main];
+      }
+    },
+    dispatch
+  };
+}
+export const actionManagerService: Service<ActionManager> = {
+  name: "actionManager",
```

Comment:
rename service to `action_manager`

Reply:
i expect the action manager service to depends on:
- `router` service => to read/update the url
- `rpc` service => to perform rpc (such as load action)
- `menus` service => to query menu info if necessary (not sure about this one)